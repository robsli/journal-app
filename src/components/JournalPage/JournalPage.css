.wrapper {
  display: flex;
  flex-direction: column;
  min-height: 100vh;
  background-color: lightgray;
}
.journal-app {
  display: flex;
  flex-direction: column;
  width: 80vw;
  /* max-height: 90vh; */
  margin: 16px auto;
  background-color: white;
  box-shadow: 0 3px 6px rgba(0,0,0,0.16), 0 3px 6px rgba(0,0,0,0.24);
}
.body {
  display: flex;
  flex-direction: row;
  padding: 0 16px;
  position: relative;
}
.body > .fa {
  border: 0;
  border-radius: 50%;
  color: rgb(97, 97, 97);
  cursor: pointer;
  font-size: 2rem;
  left: 24px;
  position: absolute;
  top: 16px;
  width: 2rem;

  transform: rotate(45deg);
  transition: all 0.1s ease;
  text-align: center;
  -webkit-text-stroke: 4px white;
}
.body > .fa.show {
  transform: rotate(180deg);
}
.body > .fa:hover {
  color: black;
}

/* Modal Styles */
.modal-overlay {
  bottom: 0;
  display: flex;
  justify-content: center;
  left: 0;
  position: fixed;
  right: 0;
  top: 0;

  background-color: rgba(0, 0, 0, 0.6);
}
.user-modal {
  align-items: center;
  bottom: 20%;
  display: flex;
  flex-direction: column;
  height: 400px;
  width: 350px;
  position: absolute;
  top: 10vw;
  
  background-color: white;
  box-shadow: 0 10px 20px rgba(0,0,0,0.24), 0 8px 8px rgba(0,0,0,0.32);
  font-family: Helvetica, Arial, sans-serif;
  outline: none;
}
.user-modal > h1 {
  margin-top: 48px;
  margin-bottom: 32px;
}
.user-modal > .input-group {
  align-items: flex-end;
  border-bottom: 1px solid black;
  display: flex;
  flex-direction: row;
  height: 24px;
  margin: 16px 0;
  width: 80%;
}
.user-modal > .input-group > i {
  align-self: center;
  margin: 0 20px 4px 12px;
}
.user-modal > .input-group > input {
  border: none;
  outline: none;
}
.modal-actions {
  display: flex;
  justify-content: space-around;
  margin-top: 32px;
  width: 80%;
}
.modal-actions > button {
  border: none;
  height: 36px;
  margin: 16px 8px;

  cursor: pointer;
  outline: none;
  transition: all 200ms ease;
}
.modal-actions > .cancel {
  flex: 1 0;
}
.modal-actions > .cancel:hover {
  text-decoration: underline;
}
.modal-actions > .login {
  flex: 3 0;

  background: rgb(80, 80, 80);
  font-size: 0.8rem;
  color: white;
}
.modal-actions > .login:hover {
  background: rgb(60, 60, 60);

  transform: scale(1.05);
}

@media (max-width: 599px) {
  .journal-app {
    margin: 0;
    width: 100%;
  }
  .body {
    flex-direction: column;
  }
}
