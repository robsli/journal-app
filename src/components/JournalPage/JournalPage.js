import React from 'react';
import EntriesApi from '../../api/entriesApi';
import Header from '../Common/Header';
import EntryForm from '../EntryForm/EntryForm';
import JournalEntries from '../JournalEntries/JournalEntries';
import SearchBar from '../SearchBar/SearchBar';
import './JournalPage.css';
import formatDate from '../Common/FormatDate';

class JournalPage extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      currentEntry: emptyEntry,
      journalEntries: [],
      searchInitiatied: false,
      searchItem: '',
      searchResults: [],
      showForm: false,
      viewEntries: false
    }
    
    this.handleChange = this.handleChange.bind(this);
    this.handleDelete = this.handleDelete.bind(this);
    this.handleSearch = this.handleSearch.bind(this);
    this.handleUpdate = this.handleUpdate.bind(this);
    this.saveEntry = this.saveEntry.bind(this);
    this.toggleView = this.toggleView.bind(this);
  }

  componentDidMount() {
    EntriesApi.getAllEntries()
      .then((response) => {
        this.setState({ journalEntries: response });
      })
  }

  handleChange(event) {
    const name = event.target.name;
    const value = event.target.value;
    
    if (name === 'searchItem') {
      this.setState(prevState => ({
        [name]: value
      }));
    } else {
      this.setState(prevState => ({
        currentEntry: Object.assign({}, prevState.currentEntry, {[name]:value})
      }));
    }
  }

  handleDelete(entryId, event) {
    EntriesApi.deleteEntry(entryId)
      .then((response) => {
        EntriesApi.getAllEntries()
          .then((response) => {
            this.setState((prevState) => ({ journalEntries: response }));
          })
      })
      .catch((err) => {
        console.log(err);
      });
  }
  
  handleSearch(event) {
    if (event.type === 'keydown' && event.keyCode === 13) {
      if (this.state.searchItem === '') {
        this.setState(prevState => ({
          searchInitiated: false,
          searchResults: []
        }));
        // console.log('You didn\'t enter anything into the search field. Can\'t search for nothing!');
      } else {
        const searchResults = Object.assign([], this.state.journalEntries).filter(entry => {
          return (entry.name.toLowerCase().includes(this.state.searchItem.toLowerCase()) || 
                  entry.notes.toLowerCase().includes(this.state.searchItem.toLowerCase()))
        })
        this.setState(prevState => ({
          searchInitiated: true,
          searchResults: searchResults
        }));
      }
    } else if (event.target.className === 'fa fa-times') {
      console.log(event.target);
      this.setState(prevState => ({
        searchItem: '',
        searchInitiated: false,
        searchResults: []
      }));
    } else if (event.type === 'change') {
      const name = event.target.name;
      const value = event.target.value;

      if (value !== '') {
        document.getElementsByName('clearSearch')[0].classList.add('clear-search');
      } else {
        document.getElementsByName('clearSearch')[0].classList.remove('clear-search');
      }
  
      this.setState(prevState => ({
        [name]: value
      }));
    }
  }

  handleUpdate(entryId, event) {
    const selectedEntry = Object.assign({}, this.state.journalEntries.filter(entry => {
      return entry._id === entryId;
    }))[0];
    this.setState(prevState => ({
      currentEntry: selectedEntry
    }));
    this.toggleView({target: {id: 'showForm'}});
  }
  
  saveEntry() {
    let entry = Object.assign({}, this.state.currentEntry, {
      date: new Date(this.state.currentEntry.date)
    });
    if (entry._id === undefined || entry._id === null) {
      EntriesApi.addEntry(entry)
        .then((response) => {
          EntriesApi.getAllEntries()
            .then((response) => {
              this.setState((prevState) => ({ currentEntry: emptyEntry, journalEntries: response }));
            })
        })
        .catch((err) => {
          console.log(err);
        });
    } else {
      EntriesApi.updateEntry(entry)
      .then((response) => {
        EntriesApi.getAllEntries()
          .then((response) => {
            this.setState((prevState) => ({ currentEntry: emptyEntry, journalEntries: response }));
          })
      })
      .catch((err) => {
        console.log(err);
      });
    }


    this.toggleView({target: {id: 'showForm'}});
  }

  toggleView(event) {
    const name = event.target.name || event.target.id;
    this.setState(prevState => ({
      [name]: !prevState[name]
    }))
    if (this.state.showForm) {
      this.setState(prevState => ({
        currentEntry: emptyEntry
      }));
    };
    if (name === 'showForm') {
      document.getElementsByClassName('form-component')[0].classList.toggle('open');
    }
  }

  render() {
    return (
      <div className='wrapper'>
        <Header />
        <div className='journal-app'>
          <SearchBar
              handleChange = { this.handleSearch }
              handleSearch = { this.handleSearch }
              searchItem = { this.state.searchItem } />
          <div className='body'>
            <JournalEntries
              formatDate = { formatDate }
              handleDelete = { this.handleDelete }
              handleUpdate = { this.handleUpdate }
              journalEntries = { this.state.searchInitiated ? this.state.searchResults : this.state.journalEntries }
              viewEntries = { this.state.viewEntries } />
            <EntryForm
              currentEntry = { this.state.currentEntry }
              formatDate = { formatDate }
              handleChange = { this.handleChange }
              saveEntry = { this.saveEntry }
              showForm = { this.state.showForm }
              toggleView = { this.toggleView } />
          </div>
        </div>
      </div>
    );
  }
}

const emptyEntry = {
  date: new Date(),
  name: '',
  notes: ''
}

// function formatDate(date) {
//   if (date instanceof Date) {
//     const year = date.getFullYear();
//     const month = date.getMonth()+1;
//     const day = date.getDate();

//     return `${year}-${month}-${day}`
//   }
//   return null;  
// }

export default JournalPage;
