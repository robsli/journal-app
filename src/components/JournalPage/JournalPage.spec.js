import { configure, mount, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import React from 'react'
import renderer from 'react-test-renderer'

import EntriesApi from '../../api/entriesApi'
import JournalPage from './JournalPage'

configure({ adapter: new Adapter() })

jest.mock('../../api/entriesApi')

describe('<JournalPage /> component', () => {
  let initialState
  let journalPage

  beforeEach(() => {
    initialState = {
      formHidden: true,
      journalEntries: [
        { _id: 1, name: 'Entry 1', createdDate: new Date('2018-01-01'), notes: 'Entry 1 notes' },
        { _id: 2, name: 'Entry 2', createdDate: new Date('2018-02-02'), notes: 'Entry 2 notes' }
      ],
      searchQuery: '',
      selectedEntryId: ''
    }
    journalPage = shallow(<JournalPage />)
  })

  it('calls the EntriesApi correctly on load and accurately updates the state', async () => {
    // Arrange
    journalPage.setState(initialState)

    // Act
    await journalPage.instance().componentDidMount()
    journalPage.update()

    // Assert
    expect(journalPage.state()).toEqual({
      formHidden: true,
      journalEntries: [
        { _id: 1, name: 'Entry 1', createdDate: new Date('2018-01-01'), notes: 'Entry 1 notes' },
        { _id: 2, name: 'Entry 2', createdDate: new Date('2018-02-02'), notes: 'Entry 2 notes' }
      ],
      searchQuery: '',
      selectedEntryId: ''
    })
  })

  it('opens the add entry form when user clicks the "+" button', () => {
    // Arrange
    journalPage.setState(initialState)
    expect(journalPage.state().formHidden).toBe(true)

    // Act
    const addIcon = journalPage.find('.fa-times')
    addIcon.simulate('click')
    const rendered = renderer.create(journalPage)

    // Assert
    expect(journalPage.state().formHidden).toBe(false)
    expect(rendered.toJSON()).toMatchSnapshot()

  })

  it('correctly clears the searchQuery when handleClearSearch is called', () => {
    // Arrange
    initialState.searchQuery = 'test'
    journalPage.setState(initialState)

    // Act
    journalPage.instance().handleClearSearch()

    // Assert
    expect(journalPage.state().searchQuery).toBe('')
  })

  it('opens entry form and passes correct information to EntryForm component when handleUpdate is called', () => {
    // Arrange
    journalPage.setState(initialState)

    // Act
    journalPage.instance().handleUpdate(2)
    journalPage.update()
    const rendered = renderer.create(journalPage)

    // Assert
    expect(journalPage.state().selectedEntryId).toBe(2)
    expect(journalPage.state().formHidden).toBe(false)
    expect(rendered.toJSON()).toMatchSnapshot()
  })

  it('deletes the correct entry when handleDelete is called', () => {
    // TODO: this test does not actually work due to how the function is implemented.
    //       It calls EntriesApi.getAllEntries(), which then refreshes to the original mock data state
    
    // Arrange
    journalPage.setState(initialState)

    // Act
    journalPage.instance().handleDelete(2)
    journalPage.update()
    const rendered = renderer.create(journalPage)

    expect(rendered.toJSON()).toMatchSnapshot()
  })
})
