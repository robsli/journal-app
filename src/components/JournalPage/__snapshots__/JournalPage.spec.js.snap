// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<JournalPage /> component clears the selected entry details when user closes entry form 1`] = `
<div
  className="wrapper"
>
  <div
    className="header"
  >
    Journal App
  </div>
  <div
    className="journal-app"
  >
    <div
      className="search-container"
    >
      <input
        name="searchQuery"
        onChange={[Function]}
        placeholder="Enter search item here"
        size="10"
        type="text"
        value=""
      />
    </div>
    <div
      className="body"
    >
      <div
        className="entry-list-component"
      >
        <h2>
          Journal Entries
        </h2>
        <div
          className="entry-cards"
        >
          <div
            className="entry-card"
          >
            <div
              className="entry-card-heading"
            >
              <span
                className="entry-name"
              >
                Entry 1
              </span>
              <span
                className="entry-date"
              >
                Sunday, Dec 31, 2017
              </span>
              <span
                className="entry-update"
                onClick={[Function]}
              >
                update
              </span>
              <span
                className="entry-delete"
                onClick={[Function]}
              >
                delete
              </span>
            </div>
            <span
              className="entry-notes"
            >
              Entry 1 notes
            </span>
          </div>
          <div
            className="entry-card"
          >
            <div
              className="entry-card-heading"
            >
              <span
                className="entry-name"
              >
                Entry 2
              </span>
              <span
                className="entry-date"
              >
                Thursday, Feb 1, 2018
              </span>
              <span
                className="entry-update"
                onClick={[Function]}
              >
                update
              </span>
              <span
                className="entry-delete"
                onClick={[Function]}
              >
                delete
              </span>
            </div>
            <span
              className="entry-notes"
            >
              Entry 2 notes
            </span>
          </div>
        </div>
      </div>
      <i
        className="fa fa-times"
        onClick={[Function]}
      />
    </div>
  </div>
</div>
`;

exports[`<JournalPage /> component deletes the correct entry when handleDelete is called 1`] = `
<div
  className="wrapper"
>
  <div
    className="header"
  >
    Journal App
  </div>
  <div
    className="journal-app"
  >
    <div
      className="search-container"
    >
      <input
        name="searchQuery"
        onChange={[Function]}
        placeholder="Enter search item here"
        size="10"
        type="text"
        value=""
      />
    </div>
    <div
      className="body"
    >
      <div
        className="entry-list-component"
      >
        <h2>
          Journal Entries
        </h2>
        <div
          className="entry-cards"
        >
          <div
            className="entry-card"
          >
            <div
              className="entry-card-heading"
            >
              <span
                className="entry-name"
              >
                Entry 1
              </span>
              <span
                className="entry-date"
              >
                Sunday, Dec 31, 2017
              </span>
              <span
                className="entry-update"
                onClick={[Function]}
              >
                update
              </span>
              <span
                className="entry-delete"
                onClick={[Function]}
              >
                delete
              </span>
            </div>
            <span
              className="entry-notes"
            >
              Entry 1 notes
            </span>
          </div>
          <div
            className="entry-card"
          >
            <div
              className="entry-card-heading"
            >
              <span
                className="entry-name"
              >
                Entry 2
              </span>
              <span
                className="entry-date"
              >
                Thursday, Feb 1, 2018
              </span>
              <span
                className="entry-update"
                onClick={[Function]}
              >
                update
              </span>
              <span
                className="entry-delete"
                onClick={[Function]}
              >
                delete
              </span>
            </div>
            <span
              className="entry-notes"
            >
              Entry 2 notes
            </span>
          </div>
        </div>
      </div>
      <i
        className="fa fa-times"
        onClick={[Function]}
      />
    </div>
  </div>
</div>
`;

exports[`<JournalPage /> component opens entry form and passes correct information to EntryForm component when handleUpdate is called 1`] = `
<div
  className="wrapper"
>
  <div
    className="header"
  >
    Journal App
  </div>
  <div
    className="journal-app"
  >
    <div
      className="search-container"
    >
      <input
        name="searchQuery"
        onChange={[Function]}
        placeholder="Enter search item here"
        size="10"
        type="text"
        value=""
      />
    </div>
    <div
      className="body"
    >
      <div
        className="entry-list-component"
      >
        <h2>
          Journal Entries
        </h2>
        <div
          className="entry-cards"
        >
          <div
            className="entry-card"
          >
            <div
              className="entry-card-heading"
            >
              <span
                className="entry-name"
              >
                Entry 1
              </span>
              <span
                className="entry-date"
              >
                Sunday, Dec 31, 2017
              </span>
              <span
                className="entry-update"
                onClick={[Function]}
              >
                update
              </span>
              <span
                className="entry-delete"
                onClick={[Function]}
              >
                delete
              </span>
            </div>
            <span
              className="entry-notes"
            >
              Entry 1 notes
            </span>
          </div>
          <div
            className="entry-card"
          >
            <div
              className="entry-card-heading"
            >
              <span
                className="entry-name"
              >
                Entry 2
              </span>
              <span
                className="entry-date"
              >
                Thursday, Feb 1, 2018
              </span>
              <span
                className="entry-update"
                onClick={[Function]}
              >
                update
              </span>
              <span
                className="entry-delete"
                onClick={[Function]}
              >
                delete
              </span>
            </div>
            <span
              className="entry-notes"
            >
              Entry 2 notes
            </span>
          </div>
        </div>
      </div>
      <i
        className="fa fa-times show"
        onClick={[Function]}
      />
      <div
        className="form-component"
      >
        <div
          className="form-container"
        >
          <h2>
            Add Journal Entry
          </h2>
          <div
            className="input-group"
          >
            <label
              name="date-label"
            >
              Date
            </label>
            <input
              disabled={false}
              name="createdDate"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="2018-02-02"
              type="date"
              value="2018-02-02"
            />
          </div>
          <div
            className="input-group"
          >
            <label
              name="name-label"
            >
              Name
            </label>
            <input
              name="name"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Name"
              type="text"
              value="Entry 2"
            />
          </div>
          <label>
            Notes
          </label>
          <textarea
            name="notes"
            onChange={[Function]}
            value="Entry 2 notes"
          />
          <button
            onClick={[Function]}
          >
            Save
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<JournalPage /> component opens the add entry form when user clicks the "+" button 1`] = `
<div
  className="wrapper"
>
  <div
    className="header"
  >
    Journal App
  </div>
  <div
    className="journal-app"
  >
    <div
      className="search-container"
    >
      <input
        name="searchQuery"
        onChange={[Function]}
        placeholder="Enter search item here"
        size="10"
        type="text"
        value=""
      />
    </div>
    <div
      className="body"
    >
      <div
        className="entry-list-component"
      >
        <h2>
          Journal Entries
        </h2>
        <div
          className="entry-cards"
        >
          <div
            className="entry-card"
          >
            <div
              className="entry-card-heading"
            >
              <span
                className="entry-name"
              >
                Entry 1
              </span>
              <span
                className="entry-date"
              >
                Sunday, Dec 31, 2017
              </span>
              <span
                className="entry-update"
                onClick={[Function]}
              >
                update
              </span>
              <span
                className="entry-delete"
                onClick={[Function]}
              >
                delete
              </span>
            </div>
            <span
              className="entry-notes"
            >
              Entry 1 notes
            </span>
          </div>
          <div
            className="entry-card"
          >
            <div
              className="entry-card-heading"
            >
              <span
                className="entry-name"
              >
                Entry 2
              </span>
              <span
                className="entry-date"
              >
                Thursday, Feb 1, 2018
              </span>
              <span
                className="entry-update"
                onClick={[Function]}
              >
                update
              </span>
              <span
                className="entry-delete"
                onClick={[Function]}
              >
                delete
              </span>
            </div>
            <span
              className="entry-notes"
            >
              Entry 2 notes
            </span>
          </div>
        </div>
      </div>
      <i
        className="fa fa-times show"
        onClick={[Function]}
      />
      <div
        className="form-component"
      >
        <div
          className="form-container"
        >
          <h2>
            Add Journal Entry
          </h2>
          <div
            className="input-group"
          >
            <label
              name="date-label"
            >
              Date
            </label>
            <input
              disabled={false}
              name="createdDate"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="2018-05-14"
              type="date"
              value="2018-05-14"
            />
          </div>
          <div
            className="input-group"
          >
            <label
              name="name-label"
            >
              Name
            </label>
            <input
              name="name"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Name"
              type="text"
              value=""
            />
          </div>
          <label>
            Notes
          </label>
          <textarea
            name="notes"
            onChange={[Function]}
            value=""
          />
          <button
            onClick={[Function]}
          >
            Save
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
